cmd/notify-mailer/main.go

```diff
@@ -130,68 +140,68 @@ func (m *mailer) run() error {
 
 	startTime := m.clk.Now()
 
-	for i, dest := range destinations {
-		m.printStatus(dest, i+1, len(destinations), startTime)
-		if strings.TrimSpace(dest) == "" {
+	sortedAddresses := sortAddresses(addressesToRecipients)
+
+	var sent int
+	for i, address := range sortedAddresses {
+		if !m.targetRange.includes(address) {
+			m.log.Debugf("skipping %q: out of target range")
 			continue
 		}
-		err := m.mailer.SendMail([]string{dest}, m.subject, m.emailTemplate)
+		recipients := addressesToRecipients[address]
+		i++
+		m.printStatus(address, i, len(addressesToRecipients), startTime)
+		var mailBody bytes.Buffer
+		err = m.emailTemplate.Execute(&mailBody, recipients)
+		if mailBody.Len() == 0 {
```

Comment:
I think this check would be better after the `if err != nil` check. I suspect that `mailBody` might end up being nil when a non-nil `err` is returned and we would probably want the non-nil err to be returned to the caller instead of the empty body error. That check is more useful in the case where `err == nil` and we want to make sure the template didn't render whack. Sorry I didn't catch that last pass.